### DISCUSSION TITLE
Need intellisense include path different from source include path

### DISCUSSION THREAD

**Discussion by mcriley821** [microsoft/vscode-cpptools#9927] [Q&A]
So I'm trying to eliminate include issues on a very large project. There's a build system setup with make that works correctly, but intellisense is having issues finding included files (rightfully so).

My project is setup similar to:

```
workspace/
├── Project_1_Module
│   └── src
│       └── example.h
└── Project_2_Module
    └── src
        └── main.cpp
```

In main.cpp:

```
    #include <Project_1_Module/example.h>
```

I can't change the include in main.cpp because it's actually a lot more complicated and spread throughout the entire project. There are a ton (literally more than 2k) of "modules" that are setup in this way,  but this is a minimum example explaining the issue.

Is there any way to achieve this behavior? If I need a custom extension, could someone help me with that?

Edit: changed the project hierarchy to similar output of `tree` for readability

**Comment by bobbrow** [comment]
The default configuration should be able to handle a project like this.  Am I correct in assuming that `<root project>` is the folder that you open in VS Code (a.k.a. the "workspace folder")?

Can you run the `C/C++: Log Diagnostics` command while main.cpp is open and showing squiggles and then share the output with us?  That should give us more information about how the extension is configuring itself for your source files.

Thanks!

**Reply by mcriley821** [reply]
Will do this ASAP thank you

**Reply by mcriley821** [reply]
```
-------- Diagnostics - 9/29/2022, 6:43:22 AM
Version: 1.12.4
Current Configuration:
{
    "name": "Linux",
    "includePath": [
        "${workspaceFolder}/**"
    ],
    "defines": [],
    "compilerPath": "/usr/bin/gcc",
    "cStandard": "gnu17",
    "cppStandard": "gnu++14",
    "intelliSenseMode": "linux-gcc-x64",
    "configurationProvider": "ms-vscode.makefile-tools",
    "compilerPathIsExplicit": true,
    "cStandardIsExplicit": true,
    "cppStandardIsExplicit": true,
    "intelliSenseModeIsExplicit": true,
    "mergeConfigurations": false,
    "browse": {
        "path": [
            "${workspaceFolder}/**"
        ],
        "limitSymbolsToIncludedHeaders": true
    }
}
Custom browse configuration:
{
    "browsePath": []
}
Translation Unit Mappings:
[ <workspace_folder>/<Project_2_Module>/src/main.cpp ]:
    <workspace_folder>/<Project_2_Module>/src/main.cpp
Translation Unit Configurations:
[ <workspace_folder>/<Project_2_Module>/src/main.cpp ]:
    Process ID: 523329
    Memory Usage: 385 MB
    Compiler Path: /usr/bin/gcc
    Includes:
        /usr/include/c++/9
        /usr/include/x86_64-linux-gnu/c++/9
        /usr/include/c++/9/backward
        /usr/lib/gcc/x86_64-linux-gnu/9/include
        /usr/local/include
        /usr/include/x86_64-linux-gnu
        /usr/include
    Standard Version: c++14
    IntelliSense Mode: linux-gcc-x64
    Other Flags:
        --g++
        --gnu_version=90400
Total Memory Usage: 385 MB
```

**Reply by bobbrow** [reply]
The issue is possibly this line in your `c_cpp_properties.json`:
```
  "configurationProvider": "ms-vscode.makefile-tools",
```

Unless you have a top-level Makefile in the root of your workspace that connects all the projects, that extension won't be able to serve the configurations for all of your modules (yet).

Can you also confirm that `<workspace_folder>` is the folder that you opened in VS Code?

**Reply by mcriley821** [reply]
Okay, that makes sense. Is there a projected date that you know of for makefile-tools to do this? 

And sorry, yes, <workspace_folder> is indeed the folder that is opened in VSCode

**Reply by bobbrow** [reply]
Unfortunately, I don't have a date for that yet.  I don't think we have multi-project workspaces as a feature ask yet.  Would you mind opening an issue requesting it in that repository?

**Reply by sean-mcmanus** [reply]
Are you referring to https://github.com/microsoft/vscode-makefile-tools/issues/90 ?

**Reply by bobbrow** [reply]
No, multi-root is different from multi-project (e.g. multiple projects in a single folder).
